lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,28 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
```

Comment:
> Yeah, that seems distinctly possible. I doubt there's a way to get a test to show this, but I could move it to initialize.

I think you could test it using something like:

``` ruby
memoized_replacement = Class.new(RSpec::Core::MemoizedHelpers::Memoized) do
  def initialize(*)
    force_another_thread_to_run_using_your_thread_controlling_thingy
    super
  end
end

stub_const("RSpec::Core::MemoizedHelpers::Memoized", memoized_replacement)
```

Basically, we replace `RSpec::Core::MemoizedHelpers::Memoized` with a subclass that, on initialization, forces control to another thread using your tooling for that.  That way, the thread switches between when `||=` in one thread started evaluating the right hand expression, but before that expression has completed in order to finish the assignment.

> The other possibility is to set it as soon as we know we need it, which would work with `let`, but I think there's other use cases which we might not know until the test has begun executing, at which point it could be too late.

What use cases can you think of?  (I can't think of any).  I like, if possible, to avoid the cost of new features/improvements for cases where the user isn't using a particular feature, and while it's not expensive per-se to instantiate `__memoized` when it's not used, I spent a lot of effort in 3.2 cutting unnecessary object allocations (and succeeded in cutting out about 30%!) and moving this to `initialize` would add 4 or 5 extra allocations per example.

So unless we know of any specific problems I like the idea of only instantiating `Memoized` if it's going to be needed.